include/mata/utils/ord-vector.hh

```diff
@@ -155,52 +162,64 @@ public:
-        }
-
-        while (first < last)
-        {	// while the pointers do not overlap
-            size_t middle = first + (last - first) / 2;
-            if (vec_[middle] == x)
-            {	// in case we found x
-                return;
-            }
-            else if (vec_[middle] < x)
-            {	// in case middle is less than x
-                first = middle + 1;
-            }
-            else
-            {	// in case middle is greater than x
-                last = middle;
-            }
-        }
-
-        vec_.resize(vec_.size() + 1);
-        std::copy_backward(vec_.begin() + static_cast<long>(first), vec_.end() - 1, vec_.end());
-
-        // insert the new element
-        vec_[first] = x;
+        // Someone else: Or we can use lower_bound like this, and remove all these comments and code? Is there a problem (except the one additional comparison)?
+        // Tomas Fiedor: article about better binary search here https://mhdm.dev/posts/sb_lower_bound/.
+        auto pos  = std::lower_bound(vec_.begin(), vec_.end(), x);
+        if (pos == vec_.end() || *pos != x)
+            vec_.insert(pos,x);
+
+        //TODO: measure, if there is not benefit to this custom version, remove
+        //size_t first = 0;
+        //size_t last = vec_.size();
+
+        //if ((last != 0) && (vec_.back() < x))
+        //{	// for the case which would be prevalent
+        //    // that is, the added thing can is larger than the largest thing and can be just bushed back
+        //    vec_.push_back(x);
+        //    return;
+        //}
+
+        //while (first < last)
+        //{	// while the pointers do not overlap
+        //    size_t middle = first + (last - first) / 2;
+        //    if (vec_[middle] == x)
+        //    {	// in case we found x
+        //        return;
+        //    }
+        //    else if (vec_[middle] < x)
+        //    {	// in case middle is less than x
+        //        first = middle + 1;
+        //    }
+        //    else
+        //    {	// in case middle is greater than x
+        //        last = middle;
+        //    }
+        //}
+
+        //vec_.resize(vec_.size() + 1);
+        //std::copy_backward(vec_.begin() + static_cast<long>(first), vec_.end() - 1, vec_.end());
+
+        //// insert the new element
+        //vec_[first] = x;
 
         assert(vectorIsSorted());
     }
 
+    //TODO: it is a little weird that insert is actually doing union, and we have other function called uni and unite
     virtual void insert(const OrdVector& vec) {
+        static OrdVector tmp{};
         assert(vectorIsSorted());
         assert(vec.vectorIsSorted());
+        tmp.clear();
 
-        // TODO: sometimes it is not efficient to copy both vectors to create a new one
-        vec_ = OrdVector::Union(*this, vec).vec_;
+        unite(*this,vec,tmp);
+        //Is this efficient? We want to use the existing memory of vec_ for a copy of tmp.vec_ if possible.
+        vec_ = tmp.vec_;
```

Comment:
I did some research and testing. It should do what we intend to do here, and we can remove the comment.

Alternatively, we can think whether using `std::vector::swap()` or (with normal local variable, not static variable) `std::move()` is better for our use cases. But the current implementation might be good enough.

Reply:
What does swap do? 
How did you test?

Reply:
[Swap](https://en.cppreference.com/w/cpp/container/vector/swap) swaps members of vectors between each other.

I wrote some minimal examples in [Godbolt Compiler explorer](https://godbolt.org/) where I looked at addresses of data. The address of `vec_` was correctly set to `tmp.vec_`.

Reply:
Note about `swap()` was added to the code. We will keep the assignment for now.